File: examples/chat/src/main/java/example/graphql/GraphQLFactory.java
/*
 * Copyright 2017-2020 original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package example.graphql;

import graphql.GraphQL;
import graphql.scalars.ExtendedScalars;
import graphql.schema.GraphQLSchema;
import graphql.schema.idl.RuntimeWiring;
import graphql.schema.idl.SchemaGenerator;
import graphql.schema.idl.SchemaParser;
import graphql.schema.idl.TypeDefinitionRegistry;
import io.micronaut.context.annotation.Bean;
import io.micronaut.context.annotation.Factory;
import io.micronaut.core.io.ResourceResolver;

import javax.inject.Singleton;
import java.io.BufferedReader;
import java.io.InputStreamReader;

/**
 * @author Gerard Klijs
 */
@Factory
@SuppressWarnings("Duplicates")
public class GraphQLFactory {

    @Bean
    @Singleton
    public GraphQL graphQL(ResourceResolver resourceResolver,
            MessagesDataFetcher messagesDataFetcher,
            ChatDataFetcher chatDataFetcher,
            StreamDataFetcher streamDataFetcher) {

        SchemaParser schemaParser = new SchemaParser();
        SchemaGenerator schemaGenerator = new SchemaGenerator();

        // Parse the schema.
        TypeDefinitionRegistry typeRegistry = new TypeDefinitionRegistry();

        resourceResolver
                .getResourceAsStream("classpath:schema.graphqls")
                .ifPresent(s -> typeRegistry.merge(schemaParser.parse(new BufferedReader(new InputStreamReader(s)))));

        // Create the runtime wiring.
        RuntimeWiring runtimeWiring = RuntimeWiring
                .newRuntimeWiring()
                .scalar(ExtendedScalars.DateTime)
                .type("QueryRoot", typeWiring -> typeWiring
                        .dataFetcher("messages", messagesDataFetcher))
                .type("MutationRoot", typeWiring -> typeWiring
                        .dataFetcher("chat", chatDataFetcher))
                .type("SubscriptionRoot", typeWiring -> typeWiring
                        .dataFetcher("stream", streamDataFetcher))
                .build();

        // Create the executable schema.
        GraphQLSchema graphQLSchema = schemaGenerator.makeExecutableSchema(typeRegistry, runtimeWiring);

        // Return the GraphQL bean.
        return GraphQL.newGraphQL(graphQLSchema).build();
    }
}

File: docs-examples/hello-world-python/src/test/python/micronaut/graphql/docs/GraphQLFactory.py
# tag::imports[]
from graphql import GraphQL
from graphql.schema import GraphQLSchema
from graphql.schema.idl import RuntimeWiring, SchemaGenerator, SchemaParser, TypeDefinitionRegistry
from jakarta.inject import Singleton
from java.io import BufferedReader, InputStreamReader
from micronaut.context.annotation import Bean, Factory
from micronaut.core.io import ResourceResolver

from .HelloDataFetcher import HelloDataFetcher
# end::imports[]


# tag::clazz[]
@Factory  # <1>
class GraphQLFactory:

    @Bean
    @Singleton
    def graph_ql(self, resource_resolver: ResourceResolver, hello_data_fetcher: HelloDataFetcher) -> GraphQL:  # <2>

        schema_parser = SchemaParser()
        schema_generator = SchemaGenerator()

        # Parse the schema.
        type_registry = TypeDefinitionRegistry()
        type_registry.merge(schema_parser.parse(BufferedReader(InputStreamReader(
            resource_resolver.getResourceAsStream("classpath:schema.graphqls").get()))))

        # Create the runtime wiring.
        runtime_wiring = (RuntimeWiring.newRuntimeWiring()
                          .type("Query", lambda type_wiring: type_wiring
                                .dataFetcher("hello", hello_data_fetcher))
                          .build())

        # Create the executable schema.
        graphql_schema: GraphQLSchema = schema_generator.makeExecutableSchema(type_registry, runtime_wiring)

        # Return the GraphQL bean.
        return GraphQL.newGraphQL(graphql_schema).build()
# end::clazz[]
